code/datums/ai/_ai_controller.dm

```diff
@@ -336,3 +351,235 @@ multiple modular subtrees with behaviors
+ * More dangerous / easier to override values, only use when you want to use a lazylist
+ *
+ * * key - A blackboard key, with its value set to a list
+ * * thing - a value which becomes the inner list value's key
+ * * value - what to set the inner list's value to
+ */
+/datum/ai_controller/proc/add_blackboard_key_assoc_lazylist(key, thing, value)
+	LAZYINITLIST(blackboard[key])
+	TRACK_AI_DATUM_TARGET(thing, key)
+	blackboard[key][thing] += value
+
+/**
+ * Clears the passed key, resetting it to null
+ *
+ * Not intended for use with list keys - use [proc/remove_thing_from_blackboard_key] if you are removing a value from a list at a key
+ *
+ * * key - A blackboard key
+ */
+/datum/ai_controller/proc/clear_blackboard_key(key)
+	CLEAR_AI_DATUM_TARGET(blackboard[key], key)
+	blackboard[key] = null
+
+/**
+ * Remove the passed thing from the associated blackboard key
+ *
+ * Intended for use with lists, if you're just clearing a reference from a key use [proc/clear_blackboard_key]
+ *
+ * * key - A blackboard key
+ * * thing - a value to set the blackboard key to.
+ */
+/datum/ai_controller/proc/remove_thing_from_blackboard_key(key, thing)
+	var/associated_value = blackboard[key]
+	if(thing == associated_value)
+		stack_trace("remove_thing_from_blackboard_key was called un-necessarily in a situation where clear_blackboard_key would suffice. ")
+		clear_blackboard_key(key)
+		return
+
+	if(!islist(associated_value))
+		CRASH("remove_thing_from_blackboard_key called with an invalid \"thing\" argument ([thing]). \
+			(The associated value of the passed key is not a list and is also not the passed thing, meaning it is clearing an unintended value.)")
+
+	for(var/inner_key in associated_value)
+		if(inner_key == thing)
+			// flat list
+			CLEAR_AI_DATUM_TARGET(thing, key)
+			associated_value -= thing
+			return
+
+		else if(associated_value[inner_key] == thing)
+			// assoc list
+			CLEAR_AI_DATUM_TARGET(thing, key)
+			associated_value -= inner_key
+			return
+
+	CRASH("remove_thing_from_blackboard_key called with an invalid \"thing\" argument ([thing]). \
+		(The passed value is not tracked in the passed list.)")
+
+/// Signal proc to go through every key and remove the datum from all keys it finds
+/// Works with keys that associate to lists, but no deeper (currently)
+/datum/ai_controller/proc/sig_remove_from_blackboard(datum/source)
+	SIGNAL_HANDLER
+
+	for(var/key in blackboard)
+		var/inner_value = blackboard[key]
+		if(blackboard[key] == source)
+			CLEAR_AI_DATUM_TARGET(blackboard[key], key)
+			blackboard[key] = null
+
+		else if(islist(inner_value))
+			for(var/inner_key in inner_value)
+				if(inner_key == source)
+					// flat list
+					CLEAR_AI_DATUM_TARGET(source, key)
+					inner_value -= source
+				else if(inner_value[inner_key] == source)
+					// assoc list
+					CLEAR_AI_DATUM_TARGET(source, key)
+					inner_value -= inner_key
```

Comment:
so clearing anything from the blackboard requires walking all possible instances. That's ok
Do we expect this to grow to the point where this could be a performance concern, like what happen with connect_loc?